src/components/charts/HomepageCustomCharts/StackedBarChart/index.jsx

```diff
@@ -0,0 +1,193 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      datasets: null,
+      valuesSet: null,
+    };
+  }
+
+  componentDidMount() {
+    const typeName = this.props.dataType;
+    const xAxisProp = this.props.xAxisProp;
+    const yAxisProp = this.props.yAxisProp;
+    this.downloadData(typeName, xAxisProp, yAxisProp)
+      .then(({ datasets, valuesSet }) => {
+        this.setState({ datasets, valuesSet });
+      });
+  }
+
+  async downloadData(typeName, xAxisProp, yAxisProp) {
+    try {
+      const query = `
+        query ($nestedAggFields: JSON, $filter: JSON) {
+          _aggregation {
+            ${typeName}(nestedAggFields: $nestedAggFields, filter: $filter) {
+              ${xAxisProp} {
+                histogram {
+                  key
+                  count
+                  termsFields {
+                    field
+                    terms {
+                      key
+                      count
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }`;
+      const variables = {
+        nestedAggFields: {
+          termsFields: [
+            yAxisProp,
+          ],
+        },
+        filter: {},
+      };
+      if (typeof this.props.constrains !== 'undefined') {
+        variables.filter = {
+          '=': {
+            ...this.props.constrains,
+          },
+        };
+      }
+      const body = { query, variables };
+      const res = await fetchWithCreds({
+        path: guppyGraphQLUrl,
+        method: 'POST',
+        body: JSON.stringify(body),
+      });
+
+      const yAxisValueSet = new Set(); // a set of y axis prop values
+      const tempDatasets = {}; // a temp dataset, tempDatasets[xAxis][yAxis] = count
+      // eslint-disable-next-line no-underscore-dangle
+      res.data.data._aggregation[typeName][xAxisProp].histogram.forEach(({ key, termsFields }) => {
+        const xAxisValue = key;
+  
+        // validate the response result format is correct
+        if (!(termsFields.length === 1 && termsFields[0].field === yAxisProp)) {
+          throw Error('Error when parsing chart data');
+        }
+        if (typeof tempDatasets[xAxisValue] === 'undefined') {
+          tempDatasets[xAxisValue] = {};
+        }
+  
+        // assign value to tempDatasets, also add y value to yAxisValueSet
+        termsFields[0].terms.forEach((d) => {
+          const yAxisValue = d.key;
+          yAxisValueSet.add(yAxisValue);
+          tempDatasets[xAxisValue][yAxisValue] = d.count;
+        });
+      });
+  
+      // need a map to remember y value => index
+      const yAxisValueToIndex = Array.from(yAxisValueSet)
+        .reduce((acc, cur, i) => { acc[cur] = i; return acc; }, {});
+  
+      // assign counts to datasets, datasets[xAxis] is an array of counts
+      const datasets = Object.keys(tempDatasets).map((xAxisValue) => {
+        const dataset = Array(yAxisValueSet.size);
```

Comment:
(This is something I don't know about JS) Can you call the Array constructor without new, like `new Array(arraySize)`?